### PR TITLE
fix(run): cancel sibling work when a concurrent run-path task fails

### PR DESCRIPTION
**Pull request by PranavMishra28**
### Summary

[#4005](https://github.com/openai/openai-agents-python/pull/4005) established that a failing branch of a concurrent fan-out must not leave its siblings running, and shipped `gather_with_cancel` (`util/_asyncio_tasks.py`) for it. That fix was applied to enablement checks. The run path still used bare `asyncio.gather`, which propagates the first exception while the siblings keep executing, detached, after the run has already unwound.

The sharpest instance is `_execute_tool_plan` (`run_internal/tool_planning.py`), which gathers six tool-execution categories with no surrounding `try`:

```python
) = await asyncio.gather(
    execute_function_tool_calls(...),
    execute_computer_actions(...),
    execute_custom_tool_calls(...),
    execute_shell_calls(...),
    execute_apply_patch_calls(...),
    execute_local_shell_calls(...),
)
```

If `execute_computer_actions` raises — e.g. `UserError("Computer tool safety check was not acknowledged")` — the shell and apply-patch arms are not cancelled, so commands run and patches are written to disk after the run has failed. It also bypasses `_FunctionToolBatchExecutor._raise_failure_after_draining_siblings`, which exists to prevent exactly this one level down, because the failure originates in a sibling gather arm rather than inside the batch.

The inconsistency is concrete for one specific pair of user-supplied callables: `realtime/openai_realtime.py` gathers `agent.get_system_prompt(...)` with `gather_with_cancel`, while both run-path turn functions (`run_single_turn` and `run_single_turn_streamed`) gathered the same call with a bare `asyncio.gather`. Separately, `run.py` hand-rolls cancel-and-drain inline for the guardrail/model race, with a comment describing the invariant. So the rule is applied in three places in the codebase and is absent from `run_internal/`.

This switches the 25 `run_internal/` gathers that lack `return_exceptions=True` to the existing helper — the tool-execution fan-out, the two prompt-resolution pairs, and the lifecycle-hook pairs. Mechanical call-site change, no new abstraction, no behavior change on the success path. `execute_mcp_approval_requests` wraps the result in `list()` because `gather_with_cancel` returns a tuple and the function declares `-> list[RunItem]`.

**Deliberately unchanged:** `run.py`'s guardrail/model race (already cancels and drains, and its tripwire branch cancels conditionally — a mechanical swap would change semantics); the 7 `voice/` and `sandbox/` gathers (separate subsystems); and every `gather(..., return_exceptions=True)` call, which already drains rather than orphaning.

### Test plan

Two regression tests in `tests/test_agent_prompt.py`, one per loop, following the pattern #4005 established in `tests/test_handoff_tool.py` (three `asyncio.Event`s: `slow_started` / `slow_cancelled` / `slow_finished`).

An agent is given an async `instructions` callable that blocks forever and records whether it was cancelled, plus an async `prompt` callable that waits for the first to start and then raises. Since `get_system_prompt` and `get_prompt` are gathered concurrently, the failing prompt must cancel the in-flight instructions callable:

- `test_run_cancels_sibling_instructions_when_prompt_resolution_fails` — non-streamed loop
- `test_run_streamed_cancels_sibling_instructions_when_prompt_resolution_fails` — streamed loop

Both assert `slow_cancelled.is_set()` and `slow_finished.is_set()`. On `main` both fail with `assert False` on `slow_cancelled` — the callable is still running when the run raises. With this change both pass.

Verification (mandatory local run order from `AGENTS.md`):

```
make format     # 847 files left unchanged
make lint       # All checks passed!
make typecheck  # mypy + pyright, exit 0
make tests      # 6223 passed, 4 skipped  /  38 passed, 5 skipped (serial)
```

### Issue number

N/A — searched open and closed issues and PRs for `gather_with_cancel`, `asyncio.gather`, `noop_coroutine`, `sibling`, and `on_tool_start`; the only related hits are #4005 (merged, established the helper) and #3406 (adds an `on_tool_progress` hook, which introduces another instance of the pattern rather than changing it). No duplicate found.

### Checks

- [x] I've added new tests, if relevant
- [ ] I've run `.agents/skills/code-change-verification/scripts/run.sh`
- [x] I've confirmed all verification steps pass
- [ ] If using Codex, I've run `/review` before submitting this PR

Developed with Claude Code; reviewed and tested by Pranav before marking ready for review.

### CODE REVIEW

**Review by seratch — CHANGES_REQUESTED** [review]
Thanks for the contribution. The cancel-and-drain direction and the use of the existing helper look sound.

Before we merge this, please add one controlled cross-category regression test at the `_execute_tool_plan` boundary. Have an async shell or apply-patch operation start and block, then make another category fail, for example through a rejected computer safety check. The test should verify that the original exception is preserved, the sibling receives cancellation, and its `finally` block completes before the run returns.

The current prompt tests cover the duplicated prompt-resolution paths, but they do not lock down the side-effecting tool boundary that motivates the PR. One test at the shared planning boundary is sufficient; a per-call-site matrix is not needed.

**Comment by seratch** [comment]
also, please fix the errors of `make typecheck`

**Comment by PranavMishra28** [comment]
yep on it rn just saw the failed ci as well

**Comment by PranavMishra28** [comment]
both pushed.

typecheck was my fault, and the cause is a bit non obvious: `gather_with_cancel` only declares overloads for 2 and 3 awaitables, so the 6 arg call in `_execute_tool_plan` fell through to the variadic `*awaitables: Awaitable[T]` overload and every unpacked element came back as the union of all six result types instead of its own. that is the return type mismatch at tool_planning.py:677. added 4/5/6 overloads so it matches the shape typeshed uses for `asyncio.gather`.

for the test, `test_execute_tool_plan_cancels_sibling_category_on_failure` in test_run_step_execution.py. shell executor sets `shell_started`, blocks, records cancellation, sets `shell_finished` in its finally. the computer tool `on_safety_check` awaits `shell_started` before returning False, so the rejection is guaranteed to land while the shell arm is genuinely in flight instead of racing it. asserts the original UserError propagates and that both events are set by the time `_execute_tool_plan` returns.

it fails on main with "sibling shell command was not cancelled", so it pins the planning boundary rather than just the helper. kept it to the one test, no per call site matrix.

green locally: lint, typecheck (mypy + pyright, 835 files), and 6232 + 38 tests.

**Comment by PranavMishra28** [comment]
One scope question while you are in it, since it borders on what you asked me to trim. There are 7 remaining bare `asyncio.gather` calls without `return_exceptions=True` outside `run_internal/`: 4 under `sandbox/` and 2 under `voice/`, plus `run.py` where the guardrail/model race already hand-rolls cancel-and-drain inline. I left all of those alone deliberately to keep this to the run path. Happy to leave them, or to open a separate PR for the `sandbox/` and `voice/` ones if you think the same invariant should hold there.

File: src/agents/run_internal/tool_planning.py [hunk 1]
```diff
@@ -580,7 +580,7 @@ async def _execute_tool_plan(
             shell_results,
             apply_patch_results,
             local_shell_results,
-        ) = await asyncio.gather(
+        ) = await gather_with_cancel(
```
[the remark this answers is not in the document]
Comment (seratch) [reply src/agents/run_internal/tool_planning.py:583]: this is not yet resolved

**Comment by seratch** [comment]
Thanks for checking. Please leave those outside this PR, and no follow-up PR is needed based on the call-site count alone.

`run.py` already has intentional cancel-and-drain semantics, and the remaining bare gather in `sandbox/materialization.py` is covered by its outer cancellation-and-drain block. The other sandbox and voice call sites have different ownership and terminal-event semantics, so we should only change them if we can demonstrate a concrete failure and cover it with a focused regression test.

Let's keep this PR scoped to `run_internal/`.

**Comment by PranavMishra28** [comment]
you're right, it wasn't. i had read that finding as already covered by the sibling cancel and it isn't, so thanks for holding it. reproduced it before touching anything:

```
fn_cancelled: True fn_finished: False
```

that's a mixed plan with one function tool and one computer action whose safety check rejects once the function tool is in flight. the function tool observes cancellation but has not finished unwinding by the time `_execute_tool_plan` raises.

the mechanism is the one you flagged. `_FunctionToolBatchExecutor.execute` catches the external `CancelledError`, calls `_cancel_pending_tasks_for_parent_cancellation` (tool_execution.py:1519), and that method cancels the per-call tasks, attaches the loop-level result callbacks, and returns without awaiting them. so the arm coroutine finishes while its handlers are still running, and `gather_with_cancel`'s drain sees a completed arm and re-raises.

what made this more than a one-liner: my first attempt just awaited the cancelled tasks there, and that broke `test_parent_cancellation_does_not_wait_for_tool_cleanup`. that test pins the opposite requirement, parent cancellation must return within 0.1s while cleanup is still blocked, and it's right to. the two cases arrive identically as a `CancelledError` from outside, so there was nothing at the catch site to branch on.

so `gather_with_cancel` now says which one it is. it puts a small mutable scope in a contextvar before creating the arm tasks, so the arms inherit a reference to the same object, and sets `cancelling = True` only when the exception that ended the gather is not itself a `CancelledError`:

```python
except BaseException as error:
    scope.cancelling = not isinstance(error, asyncio.CancelledError)
```

a real sibling failure is a non-cancellation exception, so it drains. an ancestor cancelling the gather, or an arm being cancelled directly, leaves the flag False and keeps today's prompt teardown. the executor waits a bounded 0.25s and only in the sibling-failure case; the callbacks are still attached first so anything outliving the window is still reported rather than dropped.

two tests. `test_execute_tool_plan_drains_function_tools_on_sibling_failure` is the repro above and fails without the executor change. `test_sibling_failure_drain_does_not_apply_to_parent_cancellation` pins the flag directly, False while running, True once a failing sibling is the cause, False outside any gather, which is the property the existing parent-cancellation test depends on.

on the other finding in that review, the terminal-event one on `any_llm_model.py`, that's #4143's file and i'll answer it there rather than split it across two threads.

full gate green locally, both mypy and pyright, 6237 tests. and understood on scope, i've left `sandbox/` and `voice/` alone.

File: src/agents/run_internal/tool_planning.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (PranavMishra28) [reply src/agents/run_internal/tool_planning.py:583]: reproduced and fixed, details in the thread comment. short version: the executor's parent-cancellation path returned without awaiting the tasks it had just cancelled, and awaiting them unconditionally broke test_parent_cancellation_does_not_wait_for_tool_cleanup, so gather_with_cancel now marks whether it is the one cancelling and the bounded drain only runs for a genuine sibling failure.

**Comment by PranavMishra28** [comment]
one thing that needs you rather than me: the Tests workflow on both this and #4143 is sitting at `action_required`, so the new pushes have no check runs at all yet. earlier pushes on both PRs ran fine and went 9/9, so this looks like the outside-contributor approval gate re-arming rather than anything wrong with the commits. whenever you approve the runs the results should show up on their own, no push needed from me.

File: src/agents/util/_asyncio_tasks.py [hunk 2]
```diff
@@ -27,18 +59,60 @@ async def gather_with_cancel(
...
-        await asyncio.gather(*tasks, return_exceptions=True)
-        raise
+        tasks = [asyncio.ensure_future(awaitable) for awaitable in awaitables]
+        try:
+            return tuple(await asyncio.gather(*tasks))
+        except BaseException as error:
+            # Only a genuine sibling failure flips the flag. If the gather itself was
+            # cancelled, an ancestor is going away and arms should unwind promptly
+            # rather than drain.
+            scope.cancelling = not isinstance(error, asyncio.CancelledError)
```
[the remark this answers is not in the document]
Comment (PranavMishra28) [reply src/agents/util/_asyncio_tasks.py]: right, and this is a real hole rather than a nitpick. reproduced it before changing anything: with an arm that raises `CancelledError` itself, the flag came back `[False]` where it should be `[True]`, so the other arms took the prompt teardown path even though nothing upstream was going away.

it no longer infers from the exception type at all. `gather_with_cancel` waits on `FIRST_COMPLETED` and decides from which arm finished: an arm that ends badly comes back as a finished task, which is proof this frame is still alive and the cancellation originates here, whereas an ancestor cancelling us raises out of the wait itself and never produces a finished arm.

two things worth recording about the shape. `FIRST_EXCEPTION` does not work for this, a cancelled task does not satisfy it, i confirmed that by hanging a probe on it rather than assuming. and i deliberately did not use `current_task().cancelling()` as you suggested, since it is 3.11+ and the CI matrix includes 3.10.

covered by `test_arm_raising_cancellederror_counts_as_a_sibling_failure`, which fails without the change, plus `test_ancestor_cancellation_is_not_read_as_a_sibling_failure` and an ordering test pinning that results stay positional through the new wait loop.

File: src/agents/run_internal/tool_execution.py [hunk 3]
```diff
@@ -1635,13 +1636,36 @@ async def _wait_post_invoke_tasks(
             timeout_seconds=_FUNCTION_TOOL_POST_INVOKE_WAIT_SECONDS,
         )
 
-    def _cancel_pending_tasks_for_parent_cancellation(self) -> None:
-        self.teardown_cancelled_tasks.update(self.pending_tasks)
-        _cancel_function_tool_tasks(self.pending_tasks)
+    async def _cancel_pending_tasks_for_parent_cancellation(self) -> None:
+        cancelled_tasks = set(self.pending_tasks)
+        self.teardown_cancelled_tasks.update(cancelled_tasks)
+        _cancel_function_tool_tasks(cancelled_tasks)
```
[the remark this answers is not in the document]
Comment (PranavMishra28) [reply src/agents/run_internal/tool_execution.py]: correct, and it was cancelling work that had already left the handler. `test_sibling_failure_does_not_cancel_post_invoke_work` printed `started: True finished: False` before the change, with `on_tool_end` mid-flight.

it now uses the same `_partition_pending_tasks` split the in-batch sibling-failure path uses, so only genuinely in-flight handlers are cancellable and post-invoke tasks are waited on instead. applied only for sibling failure: a parent cancellation is tearing the whole run down, so there is no lifecycle left to protect and nothing waits on the outcome.

File: src/agents/run_internal/tool_planning.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (PranavMishra28) [reply src/agents/run_internal/tool_planning.py:583]: this one i would have shipped, so thanks for catching it. verified directly with an arm that catches `CancelledError` and keeps cleaning up:

```
RESULT: hung, failure never propagated
```

the drain is bounded now by `_ARM_DRAIN_SECONDS`, set above the function-tool teardown window so an arm can still finish its own bounded cleanup rather than being cut off mid-unwind. arms that outlive the window get a done callback instead of being awaited, and results of arms that finished inside it are retrieved so a teardown failure is not logged as an exception that was never retrieved.

`test_cancellation_resistant_arm_cannot_block_the_failure` uses the real drain window rather than a patched constant, so against an unbounded drain it fails by timing out instead of by an attribute error. that distinction mattered: my first version of the test patched the constant and would have "passed" for the wrong reason.

**Comment by seratch** [comment]
#4185 includes your contribution credit in the commit. Thanks again for sharing this.

**Comment by PranavMishra28** [comment]
saw #4185, thanks for landing it and for the co-author credit.

for anyone reading this thread later: the shipped version keeps the same invariant, a failing arm cancels its siblings and the batch is drained before the failure propagates, and it also carries the three follow-ups from the review round, the arm-raised `CancelledError` classification, post-invoke tasks being excluded from the cancel, and the drain being bounded so an executor that ignores cancellation can't hold the run open.

the two things I'd flag from having gone through it: `FIRST_EXCEPTION` looks like the natural primitive for this but a cancelled task doesn't satisfy it, and `current_task().cancelling()` would have been the clean discriminator except it's 3.11+ and the matrix still includes 3.10.

happy to take another one in this area whenever something comes up.
